### PR TITLE
fix: logout credentials cleanup

### LINKED ISSUE #2594: Supabase CLI: Logout Does Not Restrict Access to Previously Linked Projects
**Describe the bug**

After logging out from a Supabase CLI account and logging in with a different account, it is still possible to affect the projects linked by the previous account.

**To Reproduce**
1. Run `supabase login` with Account A.
2. Use `supabase link` to link a project.
3. Execute `supabase logout` to log out.
4. Run `supabase login` with Account B.
5. Try to perform operations on the project linked by Account A.
6. See the issue where operations on the project linked by Account A are still possible.

**Expected behavior**
When logging in with Account B after logging out from Account A, operations on projects linked by Account A should be restricted.

**System information**
- Version of OS: macOS Sonoma 14.5
- Version of CLI: 1.187.10
- Version of Node.js: v18.19.1
- Chip: Apple M2
- Model: MacBook Pro (Model Identifier: Mac14,7)
- Docker: 
  - Version: 4.30.0 (149282)
  - Engine: 26.1.1
  - Compose: v2.27.0-desktop.2
  - Credential Helper: v0.8.1
  - Kubernetes: v1.29.2

**Additional context**
If applicable, add any other context about the problem here.
- Browser: N/A
- Version of supabase-js: N/A

### PR DESCRIPTION
**Pull request by avallete**
## What kind of change does this PR introduce?

- Import the zalando/go-keyring into our codebase to extend it with a `DeleteAll` method, removing all credentials associated with a service
- Remove all "Supabase CLI" credentials on user logout

## What is the current behavior?

Linked projects credentials stay within the keychain even after logout.

Fixes #2594

## Additional context

I'm pretty sure there might be a cleaner method in go to just "extend" the "zalando/go-keyring" module, instead of importing it as a whole. I think I could do some refactoring if pointed in the right direction for that. In the meantime, this allowed me to test the new behavior on OSX.

TODO: Test on Linux and Windows

### CODE REVIEW

**Review by sweatybridge — COMMENTED** [review]
LGTM. We might refactor this code further once upstream supports DeleteAll.

File: internal/utils/credentials/store_mock.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/zalando/go-keyring"
+)
+
+type mockProvider struct {
+	mockStore map[string]map[string]string
+	mockError error
+}
+
+var keyringMock *mockProvider = &mockProvider{}
+
```
Comment (sweatybridge) [inline internal/utils/credentials/store_mock.go:13]:
```suggestion
```

Not necessary as we can always instantiate a new one in test.

File: internal/utils/credentials/store_mock.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+func MockInit() {
+	keyringMock = &mockProvider{
+		mockStore: make(map[string]map[string]string),
+	}
+	storeProvider = keyringMock
+}
+
+func MockInitWithError(err error) {
+	keyringMock = &mockProvider{mockError: err}
+}
```
Comment (sweatybridge) [inline internal/utils/credentials/store_mock.go:77]:
```suggestion
```

Don't think this is used anywhere.

File: internal/utils/credentials/store.go [hunk 3]
```diff
@@ -27,38 +39,53 @@ func Get(project string) (string, error) {
 	return val, nil
 }
 
-// Stores the password of a project and username
-func Set(project, password string) error {
-	if err := assertKeyringSupported(); err != nil {
+func Get(project string) (string, error) {
+	return storeProvider.Get(project)
+}
```
[the remark this answers is not in the document]
Comment (avallete) [reply internal/utils/credentials/store.go]: I make sense, I guess it'll impact multiples files but probably only on one or two lines so I think that's alright and wont add too much noise to the PR.

I made the refactor in a dedicated commit here: https://github.com/supabase/cli/pull/2670/commits/8d1b7d271b9a2137b4d55e636171b2efb77a5268

**Comment by avallete** [comment]
Thank you for the thorough review and valuable feedback! I've made the suggested changes.

File: internal/logout/logout_test.go [hunk 4]
```diff
@@ -33,16 +33,40 @@ func TestLogoutCommand(t *testing.T) {
 		assert.Empty(t, saved)
 	})
 
+	t.Run("removes all Supabase CLI credentials", func(t *testing.T) {
+		storeProvider := &credentials.MockProvider{
+			MockStore: map[string]map[string]string{},
+		}
```
Comment (sweatybridge) [inline internal/logout/logout_test.go:39]: How about creating this inside `credentials.MockInit()`?

File: internal/utils/credentials/store_mock.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+func MockInit(mockProvider Store) func() {
+	oldStore := StoreProvider
+	teardown := func() {
+		StoreProvider = oldStore
+	}
+	StoreProvider = mockProvider
+	return teardown
+}
```
Comment (sweatybridge) [inline internal/utils/credentials/store_mock.go:66]:
```suggestion
func MockInit() func() {
	oldStore := StoreProvider
	teardown := func() {
		StoreProvider = oldStore
	}
	StoreProvider = &credentials.MockProvider{
		MockStore: map[string]map[string]string{},
	}
	return teardown
}
```

File: internal/logout/logout_test.go (hunk 4 above)
Comment (avallete) [reply internal/logout/logout_test.go:39]: Yeah, my guess was that you might want to inject values in the store at setup time in tests.
But I guess you would use Set methods to do so. 👍

**Review by sweatybridge — APPROVED** [review]

**Pull request merged**